File: python/pacman.py
# Source of truth: shared/game_constants.json

from maze import Maze

# ═══════════════════════════════════════════════════════
# EDUCATIONAL NOTE: Object Movement
# ═══════════════════════════════════════════════════════
# Pacman moves on a discrete grid (tile-by-tile), not pixel-by-pixel.
# Each call to update() moves Pacman exactly one tile in the current
# direction. This simplifies collision detection (just compare grid
# coordinates) and keeps the movement code very short.
#
# Compare with:
#   Java (Swing) → The Java version uses the same grid approach,
#   but must declare every field type (private int gridX;) and
#   write explicit getters. Python's dynamic typing and @property
#   decorators cut the boilerplate significantly.
#
# KEY CONCEPT: Grid-based movement means position is always an
# integer tile index, never a fractional pixel coordinate.
# ═══════════════════════════════════════════════════════


class Pacman:
    """Player-controlled entity with grid-based movement and input buffering.

    Pure logic class with ZERO rendering imports. Rendering is handled
    separately by the display layer.
    """

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Input Handling
    # ═══════════════════════════════════════════════════════
    # Pacman stores TWO directions: the current one (_dir_x/y)
    # and a buffered "next" one (_next_dir_x/y). When the player
    # presses an arrow key, only the buffer is updated. On the
    # next update(), the game tries the buffered direction first.
    # This "input buffering" makes controls feel responsive --
    # the player can press a key slightly early and it still works.
    #
    # Compare with:
    #   Java (Swing) → Same pattern, but Java's KeyListener
    #   requires implementing an interface with keyPressed(),
    #   keyReleased(), keyTyped(). Python/pygame just checks
    #   event.key in a simple loop -- no interface to implement.
    #
    # KEY CONCEPT: Input buffering separates "when the player
    # presses a key" from "when the character actually turns."
    # ═══════════════════════════════════════════════════════

    def __init__(self, x: int, y: int, size: int):
        self._grid_x: int = x
        self._grid_y: int = y
        self._size: int = size
        self._dir_x: int = 0
        self._dir_y: int = 0
        self._next_dir_x: int = 0
        self._next_dir_y: int = 0

    # ------------------------------------------------------------------
    # Input buffering
    # ------------------------------------------------------------------

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Input Handling
    # ═══════════════════════════════════════════════════════
    # set_direction() is called from game_panel.py when a key
    # is pressed. It only updates the BUFFER -- the actual move
    # happens later in update(). This decouples input events
    # from the game simulation tick.
    #
    # Compare with:
    #   Java (Swing) → In Java, the KeyListener callback would
    #   call pacman.setDirection(dx, dy) -- same concept, but
    #   Java requires @Override annotations and interface
    #   boilerplate. Python just passes a method reference.
    #
    # KEY CONCEPT: Decoupling input from movement lets the game
    # loop control exactly when direction changes take effect.
    # ═══════════════════════════════════════════════════════

    def set_direction(self, dx: int, dy: int) -> None:
        """Queue the desired direction.  Actual change happens in update()."""
        self._next_dir_x = dx
        self._next_dir_y = dy

    # ------------------------------------------------------------------
    # Movement
    # ------------------------------------------------------------------

    def update(self, maze: Maze) -> None:
        """Move one tile per call.

        1. Try the buffered (next) direction first.
        2. If blocked, try continuing in the current direction.
        3. If both are blocked, stand still.
        """
        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Collision Detection
        # ═══════════════════════════════════════════════════════
        # Before moving, Pacman checks maze.is_wall() at the target
        # tile. This is "pre-movement collision detection" -- we test
        # BEFORE changing position, so Pacman never enters a wall.
        # The two-step fallback (try buffered, then current direction)
        # makes movement feel smooth even when the player is pressing
        # into a wall while sliding along a corridor.
        #
        # Compare with:
        #   Java (Swing) → Identical logic, but notice Python
        #   does not need a separate "Direction" enum or switch
        #   statement. The direction is just a (dx, dy) pair of
        #   integers -- Python's duck typing keeps it simple.
        #
        # KEY CONCEPT: Pre-movement wall checks prevent the player
        # from ever occupying an illegal tile position.
        # ═══════════════════════════════════════════════════════

        # Attempt buffered direction
        new_x = self._grid_x + self._next_dir_x
        new_y = self._grid_y + self._next_dir_y

        if not maze.is_wall(new_x, new_y):
            self._dir_x = self._next_dir_x
            self._dir_y = self._next_dir_y
            self._grid_x = new_x
            self._grid_y = new_y
        else:
            # Fall back to current direction
            new_x = self._grid_x + self._dir_x
            new_y = self._grid_y + self._dir_y
            if not maze.is_wall(new_x, new_y):
                self._grid_x = new_x
                self._grid_y = new_y

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Rendering/Drawing
    # ═══════════════════════════════════════════════════════
    # These @property methods expose grid_x, grid_y, and size
    # as read-only attributes for the rendering layer. Pacman
    # itself has NO idea how it gets drawn -- it just publishes
    # its position. The renderer (game_panel.py) converts grid
    # coordinates to pixel positions: pixel_x = grid_x * TILE_SIZE.
    #
    # Compare with:
    #   Java (Swing) → public int getGridX() { return gridX; }
    #   Python's @property is syntactic sugar that lets callers
    #   write pacman.grid_x instead of pacman.getGridX().
    #   No compilation step needed -- just save and run.
    #
    # KEY CONCEPT: Logic classes expose state via properties;
    # only the rendering layer knows about pixels and screens.
    # ═══════════════════════════════════════════════════════

    @property
    def grid_x(self) -> int:
        return self._grid_x

    @property
    def grid_y(self) -> int:
        return self._grid_y

    @property
    def size(self) -> int:
        return self._size

File: typescript/src/pacman.ts
// Source of truth: shared/game_constants.json
// Pure logic module — ZERO rendering/GUI code.

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Input Handling — Direction Buffering
// ═══════════════════════════════════════════════════════
// Pacman uses an "input buffer" pattern: when the player presses an
// arrow key, the desired direction is stored in `_nextDirX/_nextDirY`
// but NOT applied immediately. On the next update tick, the game
// checks whether that direction is walkable; if not, Pacman keeps
// moving in its current direction. This decouples input events
// (asynchronous, fired at any time) from the fixed-step game loop.
//
// Compare with:
//   Java   → KeyListener stores key state in a HashMap<Integer,Boolean>;
//            the game loop polls it each tick. Same two-phase pattern.
//   Python → pygame.key.get_pressed() returns a snapshot each frame;
//            direction is resolved in update(), not in the event handler.
//   JavaScript → Identical DOM `keydown` listener, but without TS the
//            `dx: number` parameter could silently receive a string
//            from a refactoring mistake. TS catches that at compile time.
//
// KEY CONCEPT: Separating "record intent" (setDirection) from
// "execute intent" (update) is a fundamental game-architecture
// pattern that prevents input-dependent race conditions.
// ═══════════════════════════════════════════════════════

import { Maze } from './maze';

const PACMAN_START_X = 5;
const PACMAN_START_Y = 5;

/**
 * Represents the player-controlled Pacman character.
 * Uses input buffering: setDirection() stores the desired direction,
 * and update() applies it only if the target tile is not a wall.
 */
// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Object Movement — Underscore-Prefixed Privates
// ═══════════════════════════════════════════════════════
// TypeScript convention often uses `_field` for private backing
// fields paired with a public getter (`get gridX()`). The `private`
// keyword prevents external code from writing `pacman._gridX = 99`,
// while the getter exposes a read-only view. This is a lightweight
// alternative to Java's explicit getX()/setX() boilerplate.
//
// Compare with:
//   Java   → `private int gridX;` + `public int getGridX()`. Verbose
//            but universally understood. No property-syntax shortcut.
//   Python → Uses `@property` decorator: `@property def grid_x(self)`
//            to expose a "getter" on a `_grid_x` backing attribute.
//   JavaScript → Same `get gridX()` syntax exists, but nothing stops
//            someone from assigning `pacman._gridX` since JS has no
//            `private`. TS adds compile-time enforcement.
//
// KEY CONCEPT: TypeScript's `private` + getter pattern gives you
// Java-level encapsulation with Python-level syntactic convenience.
// ═══════════════════════════════════════════════════════
export class Pacman {
    private _gridX: number;
    private _gridY: number;
    private _dirX: number;
    private _dirY: number;
    private _nextDirX: number;
    private _nextDirY: number;

    constructor() {
        this._gridX = PACMAN_START_X;
        this._gridY = PACMAN_START_Y;
        this._dirX = 0;
        this._dirY = 0;
        this._nextDirX = 0;
        this._nextDirY = 0;
    }

    /**
     * Buffers the next desired direction. The actual direction change
     * happens in update() only if the new direction is valid (no wall).
     * This input buffering makes grid-based movement feel responsive.
     */
    public setDirection(dx: number, dy: number): void {
        this._nextDirX = dx;
        this._nextDirY = dy;
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Collision Detection — Wall Collision in Movement
    // ═══════════════════════════════════════════════════════
    // Before moving, Pacman checks `maze.isWall(newX, newY)`. This
    // is "predictive" collision detection: instead of moving first
    // and then undoing an illegal move, we test the destination tile
    // BEFORE updating position. The parameter `maze: Maze` is type-
    // checked at compile time — you cannot accidentally pass a Ghost
    // or a bare object here.
    //
    // Compare with:
    //   Java   → `public void update(Maze maze)` — same nominal
    //            typing; passing a non-Maze is a compile error.
    //   Python → `def update(self, maze: Maze)` — type hint only;
    //            passing a dict at runtime causes an AttributeError
    //            later, not at the call site.
    //   JavaScript → `update(maze)` — no type info; a wrong argument
    //            causes a runtime TypeError deep inside the method.
    //
    // KEY CONCEPT: TypeScript's nominal-like structural typing
    // catches wrong-argument bugs at compile time while keeping the
    // flexibility of duck typing when interfaces match.
    // ═══════════════════════════════════════════════════════
    /**
     * Attempts to move Pacman one tile per update:
     *   1. Try the buffered direction (player's latest input)
     *   2. If blocked by a wall, try continuing in the current direction
     *   3. If both are blocked, Pacman stands still
     */
    public update(maze: Maze): void {
        // Try to move in the buffered (next) direction
        let newX: number = this._gridX + this._nextDirX;
        let newY: number = this._gridY + this._nextDirY;

        if (!maze.isWall(newX, newY)) {
            this._dirX = this._nextDirX;
            this._dirY = this._nextDirY;
            this._gridX = newX;
            this._gridY = newY;
        } else {
            // Fall back to continuing in current direction
            newX = this._gridX + this._dirX;
            newY = this._gridY + this._dirY;
            if (!maze.isWall(newX, newY)) {
                this._gridX = newX;
                this._gridY = newY;
            }
        }
    }

    /** Current grid X coordinate. */
    public get gridX(): number {
        return this._gridX;
    }

    /** Current grid Y coordinate. */
    public get gridY(): number {
        return this._gridY;
    }
}
